03/arr-to-bst.md

```diff
@@ -0,0 +1,263 @@
+* title: Write Unit Tests
+* topics: pse
+##### !question
+
+1. Use the comments provided to write at least two example input/outputs:
+    * Consider at least one nominal and one edge case.
+    * What is the expected output for the given input?
+    * You can use the examples provided in the prompt, or other examples.
+2. Write unit tests for `arr_to_bst` for the nominal and edge cases you identified in the first step.
+
+*Note: Click the **Run Tests** button to save your tests for instructor feedback. No real tests are actually run again your unit tests.*
+
+##### !end-question
+##### !placeholder
+
+```py
+# example input 1:
+# expected output 1:
+
+# example input 2:
+# expected output 2:
+
+# To assist with testing, you also have access to the following functions:
+#   self.is_bst(root) - returns whether the tree with the given root is a valid BST
+#   self.is_balanced(root) - returns whether the tree with the given root is height balanced
+
+def test_nominal_case():
+    # ^rename with meaningful test name
+    # and complete test implementation below
+    pass
+    # arrange
+
+    # act
+
+    # assert
+
+def test_edge_case():
+    # ^rename with meaningful test name
+    # and complete test implementation below
+    pass
+    
+    # arrange
+    
+    # act
+    
+    # assert
+```
+##### !end-placeholder
+
+##### !tests
+
+```py
+import unittest
+
+class TestPython1(unittest.TestCase):
+    # Below are functions used to test if the given tree is a balanced Binary Search Tree.
+
+    # Returns True if the BST provided is a valid BST.
+    def is_bst(self, root):
+        if root is None:
+            return True
+
+        left = root.left
+        if left is not None and root.val <= left.val:
+            return False
+
+        right = root.right
+        if right is not None and root.val >= right.val:
+            return False
+
+        return self.is_bst(left) and self.is_bst(right)
+
+    # Returns the height of a tree
+    def height(self, root):
+        if root is None:
+            return 0
+        
+        left_height = self.height(root.left)
+        right_height = self.height(root.right)
+
+        if left_height > right_height:
+            return left_height + 1
+        else:
+            return right_height + 1
+
+
+    # Returns True if a tree is balanced
+    def is_balanced_tree(self, root):
+        if root is None:
+            return True
+
+        left_height = self.height(root.left)
+        right_height = self.height(root.right)
+
+        if abs(left_height - right_height) > 1:
+            return False
+
+        left_check = self.is_balanced_tree(root.left)
+        right_check = self.is_balanced_tree(root.right)
+
+        return left_check and right_check
+        
+    def test_always_pass(self):
+        self.assertEqual(1,1)
+```
+
+##### !end-tests
+##### !explanation 
+
+Example tests:
+
+```python
+def test_will_return_balanced_bst_for_odd_lengthed_list(self):
```

Comment:
We should probably remove `self` in all the examples as that's also a carry over from `unittest`